src/common/iceberg.cpp

```diff
@@ -188,19 +188,32 @@ string GenerateMetaDataUrl(FileSystem &fs, const string &meta_path, string &tabl
 }
 
 
-string IcebergSnapshot::GetMetaDataPath(const string &path, FileSystem &fs, string metadata_compression_codec, string table_version = DEFAULT_VERSION_HINT_FILE, string version_format = DEFAULT_TABLE_VERSION_FORMAT) {
+string IcebergSnapshot::GetMetaDataPath(const string &path, FileSystem &fs, string metadata_compression_codec, string table_version = DEFAULT_TABLE_VERSION, string version_format = DEFAULT_TABLE_VERSION_FORMAT) {
+	string version_hint;
+	string meta_path = fs.JoinPath(path, "metadata");
 	if (StringUtil::EndsWith(path, ".json")) {
+		// We've been given a real metadata path. Nothing else to do.
 		return path;
-	}
-
-	auto meta_path = fs.JoinPath(path, "metadata");
-	string version_hint;
-		if(StringUtil::EndsWith(table_version, ".text")||StringUtil::EndsWith(table_version, ".txt")) {
-		version_hint = GetTableVersion(meta_path, fs, table_version);
-	} else {
+	} else if (!fs.DirectoryExists(meta_path)) {
+		// Make sure we have a metadata directory to look in
+		throw IOException("Cannot open \""+path+"\": Metadata directory does not exist");
+	} else if(StringUtil::EndsWith(table_version, ".text")||StringUtil::EndsWith(table_version, ".txt")) {
+		// We were given a hint filename
+		version_hint = GetTableVersionFromHint(meta_path, fs, table_version);
+		return GenerateMetaDataUrl(fs, meta_path, version_hint, metadata_compression_codec, version_format);
+	} else if(!StringUtil::StartsWith(table_version, "?")) {
+		// We were given an explicit version number
 		version_hint = table_version;
+		return GenerateMetaDataUrl(fs, meta_path, version_hint, metadata_compression_codec, version_format);
+	} else if(fs.FileExists(fs.JoinPath(meta_path, DEFAULT_VERSION_HINT_FILE))) {
+		// We're guessing, but a version-hint.text exists so we'll use that
+		version_hint = GetTableVersionFromHint(meta_path, fs, DEFAULT_VERSION_HINT_FILE);
+		return GenerateMetaDataUrl(fs, meta_path, version_hint, metadata_compression_codec, version_format);
+	} else {
+		// We need to guess from file paths
+		return GuessTableVersion(meta_path, fs, table_version, metadata_compression_codec, version_format);
```

Comment:
I've talked with Sam about this, we feel that this could be unsafe.
Potentially you could read an uncommitted version, so we agreed that enabling this behavior should probably be put behind an `unsafe_ ...` setting that is opt-in.

I plan on making this change to your branch, just to let you know 👍

Reply:
Yea so after reading [this](https://iceberg.apache.org/spec/#metastore-tables) part of the spec, and reading some discussion on Static Tables in the Iceberg repository, we concluded that unless you use the (deprecated!) [file system tables](https://iceberg.apache.org/spec/#file-system-tables) listing the version files the way its currently done in this PR would break ACID guarantees. It's probably sensible to lock this behind a flag forcing the users to be aware of this.

Reply:
I fully agree with having this behind some sort of unsafe flag. After seeing lots of continued discussion around issue #29 I wanted to see if we could provide some finality to the functionality to close the issue. I'm not sure if there is any other way to address this without using a catalog or a hint file, but I can review the spec again (thank you for linking).

Reply:
As far as I understand it @teaguesterling there is simply no "safe" way to do this in iceberg without a catalog. The way to read a static table is by supplying the full path to the metadata file of a manually picked version

Reply:
That's my understanding as well. The number of people that wanted this (and the duckdb "spirit" of just making it work) motivated me to implement this.

Reply:
Thank you again for implementing the config variable. I've moved things around a bit for (what I perceive as) increased readability.